job/job.go

```diff
@@ -101,6 +106,84 @@ func (*jobService) RemoveJobProv(ctx context.Context, job *jobTypes.Job) error {
 	return prov.DestroyJob(ctx, job)
 }
 
+type DeployOptions struct {
+	Image string
+	Kind  provisionTypes.DeployKind
+}
+
+func (o *DeployOptions) GetKind() (kind provisionTypes.DeployKind) {
+	if o.Kind != "" {
+		return o.Kind
+	}
+
+	defer func() { o.Kind = kind }()
+	if o.Image != "" {
+		return provisionTypes.DeployImage
+	}
+
+	return provisionTypes.DeployKind("")
+}
+
+// builderDeploy uses deploy-agent to push the image to tsuru's registry and deploy the job using the new pushed image
+func builderDeploy(ctx context.Context, job *jobTypes.Job, opts builder.BuildOpts) (string, error) {
+	if err := ctx.Err(); err != nil { // e.g. context deadline exceeded
+		return "", err
+	}
+
+	if job == nil {
+		return "", errors.New("job not provided")
+	}
+
+	prov, err := getProvisioner(ctx, job)
+	if err != nil {
+		return "", err
+	}
+	builder, err := builder.GetForProvisioner(prov.(provision.Provisioner))
+	if err != nil {
+		return "", err
+	}
+	return builder.BuildJob(ctx, job, opts)
+}
+
+func getRegistry(ctx context.Context, job *jobTypes.Job) (imgTypes.ImageRegistry, error) {
+	prov, err := getProvisioner(ctx, job)
+	if err != nil {
+		return "", err
+	}
+	registryProv, ok := prov.(provision.MultiRegistryProvisioner)
+	if !ok {
+		return "", nil
+	}
+	return registryProv.RegistryForObject(ctx, job)
+}
+
+func (*jobService) BaseImageName(ctx context.Context, job *jobTypes.Job) (string, error) {
+	reg, err := getRegistry(ctx, job)
+	if err != nil {
+		return "", err
+	}
+	newImage, err := image.JobBasicImageName(reg, job.Name)
+	if err != nil {
+		return "", err
+	}
+	return fmt.Sprintf("%s:latest", newImage), nil
+}
+
+func buildWithDeployAgent(ctx context.Context, job *jobTypes.Job) error {
+	// use deploy-agent to push the image to tsuru's registry
+	newImageDst, err := builderDeploy(ctx, job, builder.BuildOpts{
+		ImageID: job.Spec.Container.Image,
+	})
+	// we don't want to fail the job creation if the image push fails, yet
+	if err == nil && newImageDst != "" {
+		// deploy the job using the new pushed image
+		job.OriginalImageURL = job.Spec.Container.Image
+		job.Spec.Container.Image = newImageDst
```

Comment:
this behavior will fail terraform providers, we must store `MirrorImageURL` and maintain intact the image provided by the user. @arthurcgc